src/H5public.h

```diff
@@ -324,6 +324,23 @@ typedef uint64_t haddr_t;
 
 #ifdef H5_HAVE_PARALLEL
 #define HADDR_AS_MPI_TYPE MPI_LONG_LONG_INT
+
+/* Define a macro to safely send an hsize_t with MPI. Note
+ * that it assumes HSIZE_UNDEF == matching type's max value. */
+#if HSIZE_UNDEF == UCHAR_MAX
+#define H5_MPI_HSIZE_T MPI_UNSIGNED_CHAR
+#elif HSIZE_UNDEF == USHRT_MAX
+#define H5_MPI_HSIZE_T MPI_UNSIGNED_SHORT
+#elif HSIZE_UNDEF == UINT_MAX
+#define H5_MPI_HSIZE_T MPI_UNSIGNED
+#elif HSIZE_UNDEF == ULONG_MAX
+#define H5_MPI_HSIZE_T MPI_UNSIGNED_LONG
+#elif HSIZE_UNDEF == ULLONG_MAX
+#define H5_MPI_HSIZE_T MPI_UNSIGNED_LONG_LONG
+#else
+#error "no suitable MPI type found for hsize_t"
+#endif
+
```

Comment:
I don't think we need this level of complexity. All 64-bit data models define long long as 64 bits, which is what we've hard-coded hsize_t to be.

I would:
- Define the type to MPI_UNSIGNED_LONG_LONG
- Either name the hsize_t type HSIZE_AS_MPI_TYPE or rename the haddr_t type H5_MPI_HADDR_T so the types have name parity
- Move the values closer to the definitions of hsize_t and haddr_t, respectively so they are less likely to get out of sync

Reply:
I agree with most all of this. The only thing I would mention is that having the catch for not finding a suitable type to use for the definition is a bit more future proof than just defining the type as MPI_UNSIGNED_LONG_LONG. Though, whether or not haddr_t/hsize_t will ever be larger than 64 bits remains to be seen. The comparisons against sizes < 64 bits are probably unnecessary here however.